### PR TITLE
Include all category types in search

### PR DESCRIPTION
**Pull request by johnnymo87**
### [STAGED HERE](https://gmaps-upgrade.herokuapp.com/)

For [Expose other categories, eg 'who they help' to search interface (advanced search interface)](https://www.pivotaltracker.com/s/projects/742821/stories/53907497)

screen shot 2015-01-12 at 02 36 09

## TODO:
* [x] Make selections in dropselects "stick" after submitting search
* [ ] MAYBE? Make the dropselects searchable by pulling in [chosen](http://harvesthq.github.io/chosen/#optgroup-support).

### CODE REVIEW

**Comment by johnnymo87** [comment]
New layout:

screen shot 2015-01-03 at 23 22 29

**Comment by tansaku** [comment]
oooh - would like to show that to client - I had imagined them as in a row, but stacked like this doesn't look too bad ...

**Comment by johnnymo87** [comment]
Alright I've implemented the layout with [flex](http://css-tricks.com/snippets/css/a-guide-to-flexbox/), as recommended by @keithpops.

screen shot 2015-01-12 at 02 36 09

I've also applied flex to the left column to get the width to fill out. I can't do the same with the height, because the map doesn't load if its container doesn't have a defined height, so I'm now setting that dynamically in the JS to match this total height of the right column.

The backend for searching with multiple categories is not implemented yet. @tansaku Do you know if the client wants the inputs to be combined as ORs?

**Comment by tansaku** [comment]
great work getting this staged @johnnymo87 just pulling this out locally to take a closer look

File: app/models/organisation.rb [hunk 1]
```diff
@@ -81,9 +81,21 @@ def self.search_by_keyword(keyword)
      self.where(contains_description(keyword).or(contains_name(keyword)))
   end
 
-  def self.filter_by_category(category_id)
-    return all unless category_id.present?
-    self.joins(:categories).where(is_in_category(category_id)) #do we need to sanitize category_id?
+  def self.filter_by_categories(category_ids)
```
Comment (mtc2013) [inline app/models/organisation.rb]: We might consider rewriting this function as I've done here:
https://github.com/mtc2013/LocalSupport/commit/fab5b834d26f33cb40b09a5d62691c1c90b36425

Incidentally, I think this form would make it very easy to switch to an AND logic simply by substituting arel and for the 'or'.

Comment (johnnymo87) [inline app/models/organisation.rb]: @mtc2013 The ability to switch so easily between `and` and `or` is amazing. This inspired me to do a refactor. [Here's a diff of the specific commits](https://github.com/johnnymo87/LocalSupport/compare/fa6257c76dc27d92f19db28d9994963dbc96d836...aae15d3bb27772f8b9f84b6e6c526fb25ebd0387). Thoughts?

**Comment by tansaku** [comment]
rachel suggests ANDing the search as a better default

**Comment by mtc2013** [comment]
It turns out I was wrong about the switching to AND so simply with arel... so it seems.

**Comment by mtc2013** [comment]
@tansaku this now has 'and' logic across the categories.

**Comment by tansaku** [comment]
@johnnymo87 is the staging site updated with the AND functionality as default?

**Comment by mtc2013** [comment]
@tansaku it seems like it is, yes.

**Comment by tansaku** [comment]
great work on this @mtc2013 and @johnnymo87 and @marianmosley - apologies for the delayed feedback.  Here are some notes:

* the search method in organisations controller seems a bit long - rather than passing 7 instance variables to the view, shouldn't we just pass the parsed_params and have the view pull out the necessary components - similarly whats whos and hows could be compressed into a class method on Category, e.g.

```ruby
def Category.name_and_id_for_what_who_and_how
  { what: Category.what_they_do.pluck(:name, :id),
    who: Category.who_they_help.pluck(:name, :id),
    how: Category.how_they_help.pluck(:name, :id) }
end
```
and then a single instance variable be passed to the view, from which the necessary components could be extracted.  Or even we could have something like the following for all three:

```ruby
def Category.what_they_do_name_and_id
  Category.what_they_do.pluck(:name, :id).pluck(:name, :id)
end
```

and then in the search method pass in the Category class, e.g. @category_class = Category, and then in the view we could do @category_class.what_they_do_name_and_id ... or even we could use method_missing to check if we have 'name_and_id' appended to an existing method to automate the plucking process - but maybe that's a bridge too far :-)

* do all the new methods in organisation.rb need to be public?
* interesting to pull queries out of models and into query modules ...
* the three category steps look like they could be DRYed out:

```ruby
Then(/^the "(.*?)" category should be selected from What They Do$/) do |name|
Then(/^the "(.*?)" category should be selected from How They Help$/) do |name|
Then(/^the "(.*?)" category should be selected from Who They Help$/) do |name|
```

* similarly in nav steps

* also I wonder if we should delete our controller and view specs?

Anyhow, the AND default seems to work great and I think the client will be very pleased with that.

**Comment by mtc2013** [comment]
I've pushed some changes to address all of your comments except the ones calling for a new method to pluck the category names and ids.  I'll look at those later.

**Comment by mtc2013** [comment]
@tansaku marian and i finished implementing your proposals this evening.

File: app/models/organisation.rb (hunk 1 above)
Comment (tansaku) [inline app/models/organisation.rb]: not sure the diff highlights precisely that change ... or I'm going code blind :-)

**Comment by tansaku** [comment]
great work guys!

**Pull request merged**